File: apps/api/loomrun_api/garment_defaults.py
"""Default garment mockup registry + product-name → garment type resolver.

Maps LoomRun Design product types onto the existing TEMPLATE_2D PNG catalog
(mockup_templates/catalog.json). No AI image generation.
"""

from __future__ import annotations

import re
from typing import Any

# Product types shown in the Design UI (stable API values).
PRODUCT_TYPES: list[tuple[str, str]] = [
    ("T_SHIRT", "T-Shirt"),
    ("POLO_T_SHIRT", "Polo T-Shirt"),
    ("SHIRT", "Shirt"),
    ("PANTS", "Pants / Trousers"),
    ("SHORTS", "Shorts"),
    ("HOODIE", "Hoodie"),
    ("SWEATSHIRT", "Sweatshirt"),
    ("JACKET", "Jacket"),
    ("OTHER", "Other"),
]

PRODUCT_TYPE_KEYS = frozenset(k for k, _ in PRODUCT_TYPES)

# Maps Design product type → catalog garment_type used by mockup_compose.
# Types without dedicated assets fall back to the closest existing template.
CATALOG_GARMENT_FOR_PRODUCT: dict[str, str] = {
    "T_SHIRT": "TSHIRT",
    "POLO_T_SHIRT": "POLO",
    "SHIRT": "SHIRT",
    "PANTS": "PANT",
    "SHORTS": "SHORTS",
    "HOODIE": "HOODIE",
    "SWEATSHIRT": "HOODIE",
    "JACKET": "HOODIE",
    "OTHER": "TSHIRT",
}

DEFAULT_VIEWS = ("FRONT", "BACK", "LEFT", "RIGHT")
VIEW_LABELS = {
    "FRONT": "Front",
    "BACK": "Back",
    "LEFT": "Left",
    "RIGHT": "Right",
}

# Keyword patterns → product type (first match wins; order matters).
_DETECT_PATTERNS: list[tuple[re.Pattern[str], str]] = [
    (re.compile(r"\bpolo\b", re.I), "POLO_T_SHIRT"),
    (re.compile(r"\bhoodie\b|\bhoody\b", re.I), "HOODIE"),
    (re.compile(r"\bsweat\s*shirt\b|\bsweatshirt\b", re.I), "SWEATSHIRT"),
    (re.compile(r"\bjacket\b|\bblazer\b", re.I), "JACKET"),
    (re.compile(r"\bshorts?\b", re.I), "SHORTS"),
    (re.compile(r"\bpants?\b|\btrousers?\b|\bdenim\b|\bjeans?\b", re.I), "PANTS"),
    (re.compile(r"\bt[\s-]?shirts?\b|\btee\b", re.I), "T_SHIRT"),
    (re.compile(r"\bshirts?\b", re.I), "SHIRT"),
]

_HEX_RE = re.compile(r"^#?[0-9A-Fa-f]{6}$")


def normalize_product_type(raw: str | None) -> str | None:
    if not raw:
        return None
    key = str(raw).strip().upper().replace("-", "_").replace(" ", "_")
    if key in PRODUCT_TYPE_KEYS:
        return key
    # Accept catalog garment names too
    aliases = {
        "TSHIRT": "T_SHIRT",
        "TSHIRTS": "T_SHIRT",
        "POLO": "POLO_T_SHIRT",
        "PANT": "PANTS",
        "PANTS": "PANTS",
        "SHORT": "SHORTS",
        "SHORTS": "SHORTS",
    }
    return aliases.get(key)


def catalog_garment_type(product_type: str) -> str:
    pt = normalize_product_type(product_type) or "OTHER"
    return CATALOG_GARMENT_FOR_PRODUCT.get(pt, "TSHIRT")


def resolve_product_type_from_text(*parts: str | None) -> str | None:
    """Best-effort detect garment type from order/product naming text."""
    blob = " ".join(p for p in parts if p).strip()
    if not blob:
        return None
    for pattern, product_type in _DETECT_PATTERNS:
        if pattern.search(blob):
            return product_type
    return None


def normalize_hex_color(raw: str | None, *, default: str = "#FFFFFF") -> str:
    if not raw or not str(raw).strip():
        return default
    c = str(raw).strip()
    if not c.startswith("#"):
        c = f"#{c}"
    if not _HEX_RE.match(c):
        raise ValueError(f"Invalid colour: {raw}")
    return c.upper()


def parse_hex_rgb(color: str) -> tuple[int, int, int]:
    c = normalize_hex_color(color)
    return int(c[1:3], 16), int(c[3:5], 16), int(c[5:7], 16)


def list_product_types() -> list[dict[str, Any]]:
    return [{"value": k, "label": lab} for k, lab in PRODUCT_TYPES]


def default_views_for_product(product_type: str) -> list[dict[str, str]]:
    """Describe Front/Back/Left/Right template keys for a product type."""
    from loomrun_api import mockup_compose

    gt = catalog_garment_type(product_type)
    items: list[dict[str, str]] = []
    for view in DEFAULT_VIEWS:
        tmpl = mockup_compose.get_template(gt, view)
        if not tmpl:
            continue
        items.append(
            {
                "view": view,
                "template_key": tmpl["key"],
                "label": VIEW_LABELS.get(view, view.title()),
                "garment_type": gt,
            }
        )
    return items

File: apps/web/src/lib/garmentDefaults.ts
/**
 * Design workspace garment defaults — mirrors apps/api/.../garment_defaults.py
 * and maps onto existing TEMPLATE_2D catalog keys.
 */

export type ProductType =
  | 'T_SHIRT'
  | 'POLO_T_SHIRT'
  | 'SHIRT'
  | 'PANTS'
  | 'SHORTS'
  | 'HOODIE'
  | 'SWEATSHIRT'
  | 'JACKET'
  | 'OTHER'

export const PRODUCT_TYPE_OPTIONS: { value: ProductType; label: string }[] = [
  { value: 'T_SHIRT', label: 'T-Shirt' },
  { value: 'POLO_T_SHIRT', label: 'Polo T-Shirt' },
  { value: 'SHIRT', label: 'Shirt' },
  { value: 'PANTS', label: 'Pants / Trousers' },
  { value: 'SHORTS', label: 'Shorts' },
  { value: 'HOODIE', label: 'Hoodie' },
  { value: 'SWEATSHIRT', label: 'Sweatshirt' },
  { value: 'JACKET', label: 'Jacket' },
  { value: 'OTHER', label: 'Other' },
]

/** Catalog garment_type for TEMPLATE_2D templates. */
export const CATALOG_GARMENT: Record<ProductType, string> = {
  T_SHIRT: 'TSHIRT',
  POLO_T_SHIRT: 'POLO',
  SHIRT: 'SHIRT',
  PANTS: 'PANT',
  SHORTS: 'SHORTS',
  HOODIE: 'HOODIE',
  SWEATSHIRT: 'HOODIE',
  JACKET: 'HOODIE',
  OTHER: 'TSHIRT',
}

export const DEFAULT_GARMENT_VIEWS = ['FRONT', 'BACK', 'LEFT', 'RIGHT'] as const
export type GarmentView = (typeof DEFAULT_GARMENT_VIEWS)[number]

export const COLOR_PRESETS: { label: string; hex: string }[] = [
  { label: 'White', hex: '#FFFFFF' },
  { label: 'Black', hex: '#1A1A1A' },
  { label: 'Navy', hex: '#1E3A5F' },
  { label: 'Grey', hex: '#8B929A' },
  { label: 'Red', hex: '#C62828' },
  { label: 'Blue', hex: '#1565C0' },
  { label: 'Green', hex: '#2E7D32' },
  { label: 'Yellow', hex: '#F9A825' },
  { label: 'Beige', hex: '#D7C4A5' },
]

const DETECT: [RegExp, ProductType][] = [
  [/\bpolo\b/i, 'POLO_T_SHIRT'],
  [/\bhoodie\b|\bhoody\b/i, 'HOODIE'],
  [/\bsweat\s*shirt\b|\bsweatshirt\b/i, 'SWEATSHIRT'],
  [/\bjacket\b|\bblazer\b/i, 'JACKET'],
  [/\bshorts?\b/i, 'SHORTS'],
  [/\bpants?\b|\btrousers?\b|\bdenim\b|\bjeans?\b/i, 'PANTS'],
  [/\bt[\s-]?shirts?\b|\btee\b/i, 'T_SHIRT'],
  [/\bshirts?\b/i, 'SHIRT'],
]

export function resolveProductTypeFromText(...parts: Array<string | null | undefined>): ProductType | null {
  const blob = parts.filter(Boolean).join(' ').trim()
  if (!blob) return null
  for (const [re, type] of DETECT) {
    if (re.test(blob)) return type
  }
  return null
}

export function templateKeyFor(productType: ProductType, view: GarmentView): string {
  const g = CATALOG_GARMENT[productType]
  const map: Record<string, Record<GarmentView, string>> = {
    TSHIRT: { FRONT: 'tshirt_front', BACK: 'tshirt_back', LEFT: 'tshirt_left', RIGHT: 'tshirt_right' },
    POLO: { FRONT: 'polo_front', BACK: 'polo_back', LEFT: 'polo_left', RIGHT: 'polo_right' },
    SHIRT: { FRONT: 'shirt_front', BACK: 'shirt_back', LEFT: 'shirt_left', RIGHT: 'shirt_right' },
    HOODIE: { FRONT: 'hoodie_front', BACK: 'hoodie_back', LEFT: 'hoodie_left', RIGHT: 'hoodie_right' },
    PANT: { FRONT: 'pant_front', BACK: 'pant_back', LEFT: 'pant_left', RIGHT: 'pant_right' },
    SHORTS: { FRONT: 'shorts_front', BACK: 'shorts_back', LEFT: 'shorts_left', RIGHT: 'shorts_right' },
  }
  return map[g]?.[view] ?? map.TSHIRT[view]
}

export function normalizeHex(raw: string | null | undefined, fallback = '#FFFFFF'): string {
  if (!raw) return fallback
  let c = raw.trim()
  if (!c.startsWith('#')) c = `#${c}`
  if (!/^#[0-9A-Fa-f]{6}$/.test(c)) return fallback
  return c.toUpperCase()
}

export function presetLabel(hex: string): string | null {
  const n = normalizeHex(hex)
  return COLOR_PRESETS.find((p) => p.hex.toUpperCase() === n)?.label ?? null
}
